swarm/src/connection.rs

```diff
@@ -848,6 +920,51 @@ mod tests {
         assert_eq!(connection.handler.local_removed, vec![vec!["/foo"]]);
     }
 
+    #[test]
+    #[ignore]
+    fn repoll_with_active_protocols() {
+        fn run_benchmark(protcol_count: usize, iters: usize) {
+            // we need longer protocol names
+            macro_rules! protocols {
+                ($($name:ident)*) => {
+                    concat!(
+                        $(
+                            "/",
+                            stringify!($name),
+                            "ffffffffffffffffffffff ",
+                        )*
+                    )
+                }
+            }
+
+            const PROTOCOLS: &str = protocols!(a b c d e f g h i j k l m n o p q r s t u v);
+            let protocols = PROTOCOLS.split(' ').collect::<Vec<_>>();
+
+            let mut connection = Connection::new(
+                StreamMuxerBox::new(PendingStreamMuxer),
+                ConfigurableProtocolConnectionHandler::default(),
+                None,
+                0,
+                Duration::ZERO,
+            );
+
+            connection.handler.listen_on(&protocols[..protcol_count]);
+
+            let now = Instant::now();
+            for _ in 0..iters {
+                let _ = connection.poll_noop_waker();
+            }
+            let elapsed = now.elapsed();
+            println!("{protcol_count} {elapsed:?}");
+        }
+
+        let iters = 3_000_000;
+        run_benchmark(2, iters);
+        run_benchmark(4, iters);
+        run_benchmark(10, iters);
+        run_benchmark(20, iters);
```

Comment:
These numbers are quite small but they are realistic I'd say. However, to prove that the algorithm is becomes more efficient I'd suggest we test it with _a lot_ more protocols. Something like north of 1000 or even more. It should be much easier to measure a significant improvement for such an extreme case. If we can measure an improvement in the extreme case (perhaps even demonstrate that the algorithm no longer scales with `O(n)`?) then we win!

Reply:
```
old code
num protocols | time / op
2 249ns
4 442ns
10 1.041µs
20 2.208µs
1000 143.731µs

new code
num protocols | time / op
2 140ns
4 189ns
10 339ns
20 586ns
1000 32.706µs
``` Sadly, its still linear, its just 4-5 times faster

Reply:
> Sadly, its still linear, its just 4-5 times faster

That is still a good improvement! Well done!